languages/julia/exercises/concept/fibonacci-iterator/.docs/instructions.md

```diff
@@ -0,0 +1,86 @@
+In this exercise you're going to write an iterator that iterates through the first `n` elements of the [Fibonacci sequence](https://en.wikipedia.org/wiki/Fibonacci_number).
+
+## Fibonacci sequence
+
+The Fibonacci sequence is defined recursively.
+Each element of the sequence is the sum of the two previous elements.
+The first two elements of the sequence are set to 1.
+In maths notation:
+
+$$
+
+a_1 = a_2 = 1\\
+a_n = a_{n-1} + a_{n-2}
+
+$$
+
+For example, the third element would be $a_3 = a_1 + a_2 = 1 + 1 = 2$, the fourth would be $a_4 = a_2 + a_3 = 1 + 2 = 3$, and so on.
+
+The first elements of the sequence are $1,~1,~2,~3,~5,~8,~13,~21,~34,~55,~89,~\dots$
+
+!!! note
+    Some sources define the first elements as $a_0 = 0$ and $a_1 = 1$.
+    However in the context of this exercise, we define the sequence without a 0-th element.
```

Comment:
There are various words that you either need to define, or drop. e.g. recursion, elements. Inspired by [mathsisfun](https://www.mathsisfun.com/numbers/fibonacci-sequence.html), I would suggest the following

```suggestion
The Fibonacci sequence is a famous series of numbers, where every number is found by adding the two numbers that come before it.
The first two numbers in the sequence are both set to 1.

The start of the sequence is $1,~1,~2,~3,~5,~8,~13,~21,~34,~55,~89,~\dots$

In maths notation, the sequence is defined as:

$$

a_1 = a_2 = 1\\
a_n = a_{n-1} + a_{n-2}

$$

For example, the third element would be $a_3 = a_1 + a_2 = 1 + 1 = 2$, the fourth would be $a_4 = a_2 + a_3 = 1 + 2 = 3$, and so on.

!!! note
    Some sources define the first elements as $a_0 = 0$ and $a_1 = 1$.
    However in the context of this exercise, we define the sequence without a 0-th element.
```

Could you screenshot what this looks like with mathjax or whatever it's called pls?

Reply:
> famous series of numbers

It's not a [series](https://en.wikipedia.org/wiki/Series_(mathematics)). I don't mind leaving out or replacing jargon but the replacement shouldn't be a term that has a meaning in maths already, especially one closely related to sequence. A series would be the sum over all elements of a sequence.

> elements

That's just a normal English term, isn't it? https://www.lexico.com/definition/element 1/1.1

---

Applying some of your suggestions, what do you think of this:

```suggestion
The Fibonacci sequence is a sequence of numbers, where every element of the sequence is calculated by adding the two numbers that come before it.
You can think of a sequence as a list of numbers in a certain order.
The first two numbers of the sequence are set to 1.

The start of the sequence is $1,~1,~2,~3,~5,~8,~13,~21,~34,~55,~89,~\dots$

In maths notation:

$$

a_1 = a_2 = 1\\
a_n = a_{n-1} + a_{n-2}

$$

For example, the third element would be $a_3 = a_1 + a_2 = 1 + 1 = 2$, the fourth would be $a_4 = a_2 + a_3 = 1 + 2 = 3$, and so on.

!!! note
    Some sources define the first elements as $a_0 = 0$ and $a_1 = 1$.
    However in the context of this exercise, we define the sequence without a 0-th element.
```

---

> Could you screenshot what this looks like with mathjax or whatever it's called pls?

After applying the suggestion above:

![grafik](https://user-images.githubusercontent.com/20866761/98737310-5145bc00-23a6-11eb-9904-884b0d095781.png)

Reply:
I chose to write `$a_3 = a_1 + a_2 = 1 + 1 = 2$` etc using MathJax becaues it renders the subscript characters larger than the unicode equivalent.

Reply:
> It's not a series. I don't mind leaving out or replacing jargon but the replacement shouldn't be a term that has a meaning in maths already, especially one closely related to sequence. A series would be the sum over all elements of a sequence.

To be clear, I'm referring to [mathsisfun](https://www.mathsisfun.com/numbers/fibonacci-sequence.html) (the first hit on Google for "fibonacci sequence") when I say it's a series. And also lots of the use of that term on the [wikipeida page](https://en.wikipedia.org/wiki/Fibonacci_number). I didn't know what a series was until you explained :)

To make you despair further:
<img width="310" alt="Screenshot 2020-11-10 at 22 44 57" src="https://user-images.githubusercontent.com/286476/98742453-691d4000-23a6-11eb-843b-0d03d76d45dc.png">
<img width="310" alt="Screenshot 2020-11-10 at 22 45 00" src="https://user-images.githubusercontent.com/286476/98742456-69b5d680-23a6-11eb-93f8-921ba5551b77.png">


But yes, we should use the correct terms, so I'm happy with your change!

---

Do you think the maths definition etc adds value to the concept of iterators? I think that definition would be interesting in a Practice Exercise, but it feels like just adding more info to the student.

Maybe we could remove this bit as I don't feel it's adding anything to learning Julia:
```
For example, the third element would be $a_3 = a_1 + a_2 = 1 + 1 = 2$, the fourth would be $a_4 = a_2 + a_3 = 1 + 2 = 3$, and so on.
!!! note
    Some sources define the first elements as $a_0 = 0$ and $a_1 = 1$.
    However in the context of this exercise, we define the sequence without a $0$-th element.
```

Thoughts?

Reply:
> elements. That's just a normal English term, isn't it?

Yeah - probably. Go with it.

Reply:
> To be clear, I'm referring to mathsisfun (the first hit on Google for "fibonacci sequence") when I say it's a series. And also lots of the use of that term on the wikipeida page. I didn't know what a series was until you explained :)

I know, I saw they used it. But I'd rather avoid any potential confusion anyway.

> Maybe we could remove this bit as I don't feel it's adding anything to learning Julia:

The first part of that is an example on how to calculate it. I think that's relevant to the implementation and may help understanding it if the description "in words" is not enough. I added the note in case people already know a definition or search for it and then find other definitions (Wikipedia starts with 0 for example). It affects how you define the `length` method in step 3, so I wanted to clarify. I can remove it too though.

> Do you think the maths definition etc adds value to the concept of iterators? I think that definition would be interesting in a Practice Exercise, but it feels like just adding more info to the student.

It's literally the only non-constructed example I could think of that still has a somewhat small solution space (a lot of data structures can be iterators but they usually come with a lot of extra requirements that aren't relevant to them being iterators). I feel like a number sequence is a good balance between a somewhat realistic yet very simple context, and the Fibonacci sequence may be the best known one. It also has the advantage that the state of the iterator is more than a simple counter.

To solve it, one can ignore all the extra info. The only required context is that you calculate the next element of the iterator by adding the previous two. But imo it's nicer if there's actually some explanation. Just like string parsing exercises have the context of parsing logs when that context is not necessary to learn the language, or the basics exercise has the context of lasagna cooking.